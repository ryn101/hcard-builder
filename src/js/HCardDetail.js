import Unique from './Unique';

/**
 * Class for storing details specific to a HCard detail component
 */
class HCardDetail extends Unique {
    /**
     * Constructor
     *
     * @param label The description of the detail
     * @param {HCardProperty|[HCardProperty]} properties Either a singular or an array of
     *                                                   HCardProperty objects
     * @param propertiesSeparator The separator which should separate each property
     */
    constructor(label, properties, propertiesSeparator) {
        super('hcd');

        let propertiesArray = properties;

        if (!Array.isArray(propertiesArray)) {
            propertiesArray = [propertiesArray];
        }

        this.label = label;
        this.properties = propertiesArray;
        this.propertiesSeparator = propertiesSeparator;
    }
}

export default HCardDetail;
