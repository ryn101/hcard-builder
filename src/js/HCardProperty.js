/**
 * Configured types of hCard properties used as a lookup for value safety
 *
 * @type {{
 *    country: string,
 *    streetAddress: string,
 *    postalCode: string,
 *    locality: string,
 *    telephone: string,
 *    region: string,
 *    email: string
 *  }}
 */
const hCardPropertyType = {
    email: 'email',
    telephone: 'tel',
    streetAddress: 'street-address',
    locality: 'locality',
    region: 'region',
    postalCode: 'postal-code',
    country: 'country',
};

/**
 * Container for a HCard specific property
 */
class HCardProperty {
    /**
     * Constructor
     *
     * @param name The property name/type
     * @param value The value of the property
     */
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
}

export { hCardPropertyType as HCardPropertyType };
export default HCardProperty;
