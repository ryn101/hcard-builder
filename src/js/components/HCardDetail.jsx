import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const HCardDetail = ({ label, children, columns }) => (
    <div className={classNames('detail', `detail--columns-${columns}`)}>
        <dt className="detail__label">{ label }</dt>
        <dd className="detail__value">
            {
                children
            }
        </dd>
    </div>
);

HCardDetail.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element),
        PropTypes.element,
        PropTypes.string,
    ]).isRequired,
    label: PropTypes.string,
    columns: PropTypes.oneOf([1, 2]),
};

HCardDetail.defaultProps = {
    label: null,
    columns: 1,
};

export default HCardDetail;
