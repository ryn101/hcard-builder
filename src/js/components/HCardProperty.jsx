import React from 'react';
import PropTypes from 'prop-types';

/**
 * Wraps up the logic needed to display a hCard property. The name of the property is added
 * as a class to the component and the title is set to the specified value.
 *
 * Additionally supports property nesting, and abbreviation declaration for use with the abbr tag.
 * If child elements are provided then the value and abbreviation properties are disregarded.
 */
function HCardProperty(props) {
    const {
        name,
        value,
        abbreviation,
        children,
    } = props;

    let element;

    if (children || !abbreviation) {
        // If children are specified or no abbreviation value is specified then use the default span
        element = (<span title={value} className={name}>{ children || value }</span>);
    } else {
        // If an abbreviation is specified then create an abbr tag instead
        element = (<abbr title={value} className={name}>{abbreviation}</abbr>);
    }

    return element;
}

HCardProperty.propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    abbreviation: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.oneOfType([
            PropTypes.element,
            PropTypes.string,
        ])),
    ]),
};

HCardProperty.defaultProps = {
    value: '',
    children: null,
};

export default HCardProperty;
