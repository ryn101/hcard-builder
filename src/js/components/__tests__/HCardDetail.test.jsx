import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardDetail from '../HCardDetail';
import HCardProperty from '../HCardProperty';

configure({ adapter: new Adapter() });

it('renders the hCard detail without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <HCardDetail label="Label">
            <span />
        </HCardDetail>,
        div,
    );

    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered', () => {
    let hCardDetailNode;
    const label = 'Detail label';
    const detailChildren = [
        <HCardProperty key="1" name="property-1" />,
        <span key="2" />,
        <HCardProperty key="2" name="property-2" />,
    ];

    beforeEach(() => {
        hCardDetailNode = shallow(
            <HCardDetail label={label}>
                {
                    detailChildren
                }
            </HCardDetail>,
        );
    });

    it('renders with state correctly', () => {
        expect(hCardDetailNode).toMatchSnapshot();
    });

    it('renders the label correctly', () => {
        const labelNode = hCardDetailNode.find('.detail__label');
        expect(labelNode.text()).toEqual(label);
    });

    it('renders the children correctly', () => {
        const valueNode = hCardDetailNode.find('.detail__value');
        expect(valueNode.exists()).toBeTruthy();

        const childNodes = valueNode.children();
        expect(childNodes).toHaveLength(detailChildren.length);
    });
});
