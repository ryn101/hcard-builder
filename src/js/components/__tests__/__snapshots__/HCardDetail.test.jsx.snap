// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<div
  className="detail detail--columns-1"
>
  <dt
    className="detail__label"
  >
    Detail label
  </dt>
  <dd
    className="detail__value"
  >
    <HCardProperty
      key="1"
      name="property-1"
      value=""
    />
    <span
      key="2"
    />
    <HCardProperty
      key="2"
      name="property-2"
      value=""
    />
  </dd>
</div>
`;
